src/optimize.rs

```diff
@@ -0,0 +1,1049 @@
+//! Optimizing a Wasm program, assuming the behavior observed in the given
+//! profile.
+
+use std::collections::{HashMap, HashSet};
+
+use crate::cow_section::{borrowed, owned, CowSection};
+use crate::Profile;
+use anyhow::{bail, ensure, Result};
+use wasm_encoder::SectionId;
+use wasmparser::{Chunk, Payload};
+
+#[cfg(feature = "clap")]
+use clap::Parser;
+
+/// Optimize a Wasm program based on profiling data.
+///
+/// # Example
+///
+/// ```
+/// # fn foo() -> anyhow::Result<()> {
+/// use winliner::{Optimizer, Profile};
+///
+/// // Create and configure an optimizer.
+/// let mut optimizer = Optimizer::new();
+/// optimizer
+///     .min_total_calls(100)
+///     .min_ratio(0.99)?
+///     .max_inline_depth(5);
+///
+/// // Get the original, uninstrumented Wasm program.
+/// let wasm = std::fs::read("path/to/my.wasm")?;
+///
+/// // Get a profile for our Wasm program from somewhere. Read it from disk,
+/// // record it now in this process, etc...
+/// let profile = Profile::default();
+///
+/// // Run the optimizer with the given profile!
+/// let optimized_wasm = optimizer.optimize(&profile, &wasm)?;
+/// # Ok(()) }
+/// ```
+#[cfg_attr(feature = "clap", derive(Parser))]
+pub struct Optimizer {
+    /// The minimum number of total calls for a call site before it is
+    /// considered for winlining.
+    #[cfg_attr(feature = "clap", clap(long, default_value = "1000"))]
+    min_total_calls: u64,
+
+    /// The minimum ratio of all calls at a call site that go to a particular
+    /// callee before the callee is considered for winlining. Must be between
+    /// 0.0 and 1.0.
+    #[cfg_attr(feature = "clap", clap(long, default_value = "0.9"))]
+    min_ratio: f64,
+
+    /// The maximum inlining depth.
+    ///
+    /// This can help limit code size blowup from duplicating many function
+    /// bodies during inlining.
+    #[cfg_attr(feature = "clap", clap(long, default_value = "10"))]
+    max_inline_depth: usize,
+
+    /// Emit counters for how often our speculative inlining guesses were
+    /// correct or incorrect.
+    ///
+    /// When this option is enabled, we will add two globals for each call site
+    /// where we winlined a speculative callee:
+    ///
+    /// 1. A counter for how many times we guessed correctly.
+    ///
+    /// 2. A counter for how many times we guessed incorrectly.
+    ///
+    /// You can extract this data when using Winliner as a library with the
+    /// `Counters` type.
+    #[cfg_attr(feature = "clap", clap(long))]
+    emit_counters: bool,
```

Comment:
maybe call this feature "feedback counters" or something, to distinguish it from the initial profiling counters?